src/widget/wtrackmenu.cpp

```diff
@@ -0,0 +1,1213 @@
+        const auto indices = getTrackIndices();
+        m_pTrackInfo->loadTrack(indices.at(0));
+    } else {
+        m_pTrackInfo->loadTrack(trackPointers.at(0));
+    }
+    m_pTrackInfo->show();
+}
+
+void WTrackMenu::slotShowDlgTagFetcher() {
+    const auto trackPointers = getTrackPointers();
+    if (trackPointers.empty()) {
+        return;
+    }
+    if (m_pTrackModel) {
+        const auto indices = getTrackIndices();
+        m_pTagFetcher->loadTrack(indices.at(0));
+    } else {
+        m_pTagFetcher->loadTrack(trackPointers.at(0));
+    }
+    m_pTagFetcher->show();
+}
+
+void WTrackMenu::slotAddToAutoDJBottom() {
+    // append to auto DJ
+    addToAutoDJ(PlaylistDAO::AutoDJSendLoc::BOTTOM);
+}
+
+void WTrackMenu::slotAddToAutoDJTop() {
+    addToAutoDJ(PlaylistDAO::AutoDJSendLoc::TOP);
+}
+
+void WTrackMenu::slotAddToAutoDJReplace() {
+    addToAutoDJ(PlaylistDAO::AutoDJSendLoc::REPLACE);
+}
+
+void WTrackMenu::addToAutoDJ(PlaylistDAO::AutoDJSendLoc loc) {
+    const TrackIdList trackIds = getTrackIds();
+    if (trackIds.empty()) {
+        qWarning() << "No tracks selected for AutoDJ";
+        return;
+    }
+
+    PlaylistDAO& playlistDao = m_pTrackCollectionManager->internalCollection()->getPlaylistDAO();
+
+    // TODO(XXX): Care whether the append succeeded.
+    m_pTrackCollectionManager->unhideTracks(trackIds);
+    playlistDao.addTracksToAutoDJQueue(trackIds, loc);
+}
+
+void WTrackMenu::slotCoverInfoSelected(const CoverInfoRelative& coverInfo) {
+    const TrackPointerList trackPointers = getTrackPointers();
+    for (const auto& pTrack : trackPointers) {
+        pTrack->setCoverInfo(coverInfo);
+    }
+}
+
+void WTrackMenu::slotReloadCoverArt() {
+    const TrackPointerList trackPointers = getTrackPointers();
+    if (!trackPointers.empty()) {
+        guessTrackCoverInfoConcurrently(trackPointers);
+    }
+}
+
+void WTrackMenu::slotRemove() {
+    if (m_pTrackModel) {
+        const QModelIndexList indices = getTrackIndices();
+        if (!indices.empty()) {
+            m_pTrackModel->removeTracks(indices);
+        }
+    }
+}
+
+void WTrackMenu::slotHide() {
+    if (m_pTrackModel) {
+        const QModelIndexList indices = getTrackIndices();
+        if (!indices.empty()) {
+            m_pTrackModel->hideTracks(indices);
+        }
+    }
+}
+
+void WTrackMenu::slotUnhide() {
+    if (m_pTrackModel) {
+        const QModelIndexList indices = getTrackIndices();
+        if (!indices.empty()) {
+            m_pTrackModel->unhideTracks(indices);
+        }
+    }
+}
+
+void WTrackMenu::slotPurge() {
+    if (m_pTrackModel) {
+        const QModelIndexList indices = getTrackIndices();
+        if (!indices.empty()) {
+            m_pTrackModel->purgeTracks(indices);
+        }
+    }
+}
+
+void WTrackMenu::clearTrackSelection() {
+    m_pTrackPointerList.clear();
+    m_pTrackIndexList.clear();
+}
+
+bool WTrackMenu::modelHasCapabilities(TrackModel::CapabilitiesFlags capabilities) const {
```

Comment:
This should be a method of TrackModel, not WTrackMenu.

Reply:
Oh true. It was awkward to define it in both `WTrackTableView` and `WTrackMenu`. Will move it to the `TrackModel` class.